lib/twirp/protoc_plugin/descriptor_ext/file_descriptor_proto_ext.rb

```diff
@@ -77,4 +103,102 @@ def split_to_constants(package_or_message)
+  #     ".foo.bar.ExampleMessage.NestedMessage" => "::Foo::Bar::ExampleMessage::NestedMessage"
+  #     ".google.protobuf.Empty" => "Google::Protobuf::Empty"
+  #     ".common.bar.baz.other_type" => "::Common::Baz::OtherType"
+  #       (when type is imported from a proto file with package = "common.bar.baz"
+  #       and `option ruby_package = "Common::Baz";` specified)
+  def ruby_type_map
+    if @ruby_type_map.nil?
+      @ruby_type_map = build_ruby_type_map(self)
+    end
+
+    @ruby_type_map
+  end
+
+  # Loops through the messages in the proto file, and recurses through the messages in
+  # dependent proto files, to construct the ruby type map for all types within the file.
+  #
+  # @see [#ruby_type_map]
+  # @param proto_file [Google::Protobuf::FileDescriptorProto]
+  # @return [Hash<String, String>]
+  def build_ruby_type_map(proto_file)
+    type_map = {}
+
+    proto_file.message_type.each do |message_type|
+      add_message_type(type_map, proto_file, message_type)
+    end
+
+    proto_file.dependency_proto_files.each do |dependency_proto_file|
+      type_map.merge! build_ruby_type_map(dependency_proto_file)
+    end
+
+    type_map
+  end
+
+  # Adds the message type's key and value to the type map, recursively handling nested message
+  # types along the way.
+  #
+  # @param type_map [Hash<String, String>]
+  # @param proto_file [Google::Protobuf::FileDescriptorProto] The proto file containing the message type
+  # @param message_type [Google::Protobuf::DescriptorProto]
+  # @param parent_key [String, nil] In the recursive case, this is the parent message type key so
+  #   that the nested child type can be properly namespaced.
+  # @param parent_value [String, nil] In the recursive case, this is the parent message type value
+  #   so that the nested type can be properly namespaced.
+  # @return [void]
+  def add_message_type(type_map, proto_file, message_type, parent_key = nil, parent_value = nil)
+    key = type_map_key(proto_file, message_type, parent_key)
+    value = type_map_value(proto_file, message_type, parent_value)
+
+    type_map[key] = value
+
+    # Recurse over nested types, using the current message_type's key and value as the parent values
+    message_type.nested_type.each do |nested_type|
+      add_message_type(type_map, proto_file, nested_type, key, value)
+    end
+  end
+
+  # @param proto_file [Google::Protobuf::FileDescriptorProto] The proto file containing the message type
+  # @param message_type [Google::Protobuf::DescriptorProto]
+  # @param parent_key [String, nil] In the recursive case, this is the parent message type key so
+  #   that the nested child type can be properly namespaced.
+  # @return [String]
+  def type_map_key(proto_file, message_type, parent_key)
+    key_prefix = if !parent_key.nil?
+      "#{parent_key}."
+    elsif proto_file.package.to_s.empty?
+      "."
+    else
+      ".#{proto_file.package}."
+    end
+
+    "#{key_prefix}#{message_type.name}"
```

Comment:
I think moving the `.` into here would make it clearer.

```suggestion
    key_prefix = if !parent_key.nil?
      "#{parent_key}"
    elsif proto_file.package.to_s.empty?
      ""
    else
      ".#{proto_file.package}"
    end

    "#{key_prefix}.#{message_type.name}"
```

Reply:
I don't necessarily disagree, but I think I more strongly prefer the symmetry between the key/value helpers in terms of `"#{prefix}#{data}"`, and the extra namespace separator in the key string makes "prefix" not quite the right name anymore (maybe `key_namespace` would be better in that scenario?). I'm inclined to leave this as-is.

Reply:
Ok, then here's a simplified one without a confusing variable name ;) 

```suggestion
    if !parent_key.nil?
      "#{parent_key}.#{message_type.name}"
    elsif proto_file.package.to_s.empty?
      ".#{message_type.name}"
    else
      ".#{proto_file.package}.#{message_type.name}"
    end
```

Reply:
Ok, that's legit better. 😆

Reply:
I think I was trying to DRY `#{message_type.name}` there, which was something the original logic did before I extracted it out. 🤷

Reply:
I changed my mind after applying the suggestion, haha. I want to keep the key/value logic symmetrical because how closely those values are related to each other. Going to drop the suggestion commit.